addons/website/models/website_page.py

```diff
@@ -269,3 +285,150 @@ def action_page_debug_view(self):
+        is useful for preventing sensitive or dynamic content from being stored.
+        """
+        return (
+            request.httprequest.method == "GET"
+            and not request.params  # because the parameters are not part of the cache key
+            and request.env.user._is_public()  # only cache for unlogged user
+            and not self._get_page_info(request)['group_ids']  # do not cache elements dependent on security access
+        )
+
+    @api.model
+    def _allow_cache_insertion(self, layout):
+        """ Determines whether a page is allowed to be served from the cache
+        based on the current request, URL, or session.
+        """
+        return True
+
+    @api.model
+    def _post_process_response_from_cache(self, request: http.Request, response: http.Response) -> None:
+        """ A hook called after a response is retrieved from the cache. This
+        method allows for post-processing, such as incrementing counters or
+        modifying HTTP headers, without regenerating the entire page.
+        """
+        csrf_token = request.csrf_token(None)
+        html = response.response[0]
+        html = re.sub(r'csrf_token: "[^"]+"', f'csrf_token: {csrf_token!r}', html)
+        html = re.sub(r'name="csrf_token" value="[^"]+"', f'name="csrf_token" value={csrf_token!r}', html)
+        response.response = [html]
+
+        # used for _register_website_track
+        response._cached_view_id = self._get_page_info(request)['view_id']
+        response._cached_page = self
+
+    @api.model
+    def _get_cache_key(self, request):
+        """ Allows for supplementing the base cache key with custom components
+        (e.g., from the URL or session). This is essential for ensuring that
+        the cache serves the correct version of a page based on specific
+        parameters like user language or currency.
+        """
+        return (request.website.id, request.lang.code, request.httprequest.path, request.session.debug)
+
+    def _get_response(self, request):
+        self.ensure_one()
+        if self._allow_to_use_cache(request):
+            try:
+                response, cache_key = self._get_response_cached(request)
+            except PageCanNotBeCached as notCache:
+                if notCache.value:
+                    return notCache.value[0]
+
+            if response.time + 3600 > time.time():
+                resp = http.Response(
+                    headers=response.headers.copy(),
+                    mimetype=response.mimetype,
+                    content_type=response.content_type,
+                    status=response.status,
+                    response=[response.response[0]],
+                )
+                self._post_process_response_from_cache(request, resp)
+                return resp
+            else:
+                response = self._get_response_raw(request)
+                self._get_response_cached.__cache__.add_value(self, request, cache_value=(response, cache_key))
```

Comment:
What's that? :open_mouth:

Reply:
We're sure they'll ask the py framework :-p

We can add: "It's a tricky cache from the Python framework."